eo-phi-normalizer/src/Language/EO/Phi/Metrics.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# LANGUAGE RecordWildCards #-}
+
+{- FOURMOLU_DISABLE -}
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+
+{- FOURMOLU_ENABLE -}
+
+module Language.EO.Phi.Metrics where
+
+import Control.Lens ((+=))
+import Control.Monad (forM_)
+import Control.Monad.State (State, runState)
+import Data.Aeson (FromJSON)
+import Data.Aeson.Types (ToJSON)
+import Data.Foldable (fold)
+import Data.Function ((&))
+import Data.Generics.Labels ()
+import Data.Maybe (catMaybes)
+import Data.Traversable (forM)
+import GHC.Generics (Generic)
+import Language.EO.Phi.Rules.Common ()
+import Language.EO.Phi.Syntax.Abs
+
+data ObjectMetrics = ObjectMetrics
+  { dataless :: Int
+  , applications :: Int
+  , formations :: Int
+  , dispatches :: Int
+  }
+  deriving (Generic, Show, FromJSON, ToJSON, Eq)
+
+defaultObjectMetrics :: ObjectMetrics
+defaultObjectMetrics =
+  ObjectMetrics
+    { dataless = 0
+    , applications = 0
+    , formations = 0
+    , dispatches = 0
+    }
+
+instance Semigroup ObjectMetrics where
+  (<>) :: ObjectMetrics -> ObjectMetrics -> ObjectMetrics
+  ObjectMetrics
+    { dataless = dataless1
+    , applications = applications1
+    , formations = formations1
+    , dispatches = dispatches1
+    }
+    <> ObjectMetrics
+      { dataless = dataless2
+      , applications = applications2
+      , formations = formations2
+      , dispatches = dispatches2
+      } =
```

Comment:
nitpick: The names are already clear, no need to use record syntax. It would be shorter if you just write `ObjectMetrics dataless1 apps1 ...`

Reply:
Using field names explicitly helps with future refactoring (to avoid accidental errors) when `ObjectMetrics` changes order of fields/renames fields/adds or removes fields/etc. So I actually slightly prefer Danila's approach here.

Reply:
Addressed in [63f1829](https://github.com/objectionary/normalizer/pull/193/commits/63f1829e2ccf0e52fc1c8a7b23d809c049bb04a1)